src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -736,36 +790,81 @@ public void ExecuteOperationBlockActions(
             bool isGeneratedCode,
             CancellationToken cancellationToken)
         {
-            ExecuteBlockActionsCore<OperationBlockStartAnalyzerAction, OperationBlockAnalyzerAction, OperationAnalyzerAction, IOperation, int>(
-                operationBlockStartActions, operationBlockActions, operationBlockEndActions, analyzer,
-                declaredNode, declaredSymbol, operationBlocks, (blocks) => operations, semanticModel,
-                getKind: null, filterSpan, isGeneratedCode, cancellationToken);
+            Debug.Assert(!operationBlocks.IsEmpty);
+
+            // The actions we discover in 'addActions' and then execute in 'executeActions'.
+            var ephemeralActions = ArrayBuilder<OperationAnalyzerAction>.GetInstance();
+            ExecuteBlockActionsCore(
+                startActions,
+                actions,
+                endActions,
+                analyzer,
+                declaredNode,
+                declaredSymbol,
+                operationBlocks,
+                semanticModel,
+                filterSpan,
+                isGeneratedCode,
+                addActions: static (startAction, endActions, args, cancellationToken) =>
+                {
+                    var (@this, startActions, analyzer, declaredNode, operationBlocks, declaredSymbol, operations, semanticModel, filterSpan, isGeneratedCode, ephemeralActions) = args;
+                    var scope = new HostOperationBlockStartAnalysisScope(startAction.Analyzer);
+                    var startContext = new AnalyzerOperationBlockStartAnalysisContext(
+                        scope, operationBlocks, declaredSymbol, semanticModel.Compilation, @this.AnalyzerOptions,
```

Comment:
note: some code uses semanticModel.Compilation, and some code uses this.Compilation.  It would probably be better if we did things consistently.  I would like to move that to a followup.